### ISSUE TITLE
How to define loss function for multi-task or multi objective functions?

### ISSUE THREAD

**Issue by meshiguge** [Lasagne/Lasagne#704]
I would like to assign one training example  with two different set of labels  or two objective function. 
the output layer has two independent parts with independent target labels, and the loss objective is the addition of two cross-entropy. how to define the loss function and train function. 
My draft as follows:
# input layer
    l_in = lasagne.layers.InputLayer( input_var=input_var)
    # Hidden layer:
    l_hidden = lasagne.layers.DenseLayer( l_in,4)
    
    # Output layer:(should I write two networks as follows ????)
    # first network1 has two labels , the other has 3 labels
    network_1 = lasagne.layers.DenseLayer( l_hidden, 2, nonlinearity=lasagne.nonlinearities.softmax)
    network_2 = lasagne.layers.DenseLayer( l_hidden, 3, nonlinearity=lasagne.nonlinearities.softmax)

    # predict 
    prediction_1 = lasagne.layers.get_output(network_1)
    prediction_2 = lasagne.layers.get_output(network_2)

    # loss (should I define two loss? )
    loss_1 = lasagne.objectives.categorical_crossentropy(prediction_1, target_var_1)
    loss_2 = lasagne.objectives.categorical_crossentropy(prediction_2, target_var_2)

    or should I use this ? 
    loss = loss_1+loss_2 

    # get params 
    params_1 = lasagne.layers.get_all_params(network_1, trainable=True)
    params_2 = lasagne.layers.get_all_params(network_2, trainable=True)

    # but how to define updates? there is just one params argument 
    updates = lasagne.updates.nesterov_momentum( loss, params, learning_rate=0.01, momentum=0.9)
    # define train function , but we have two target_var ? 

    train= theano.function([input_var, target_var], loss, updates=updates)

Please give me some adivices, thank you

**Comment by f0k** [comment]
> ```python
# predict 
prediction_1 = lasagne.layers.get_output(network_1)
prediction_2 = lasagne.layers.get_output(network_2)
```

Never do this when the output layers share some inputs! Do a single `get_output` call as indicated in the documentation:
```python
prediction_1, prediction_2 = lasagne.layers.get_output([network_1, network_2])
```

> ```python
# but how to define updates? there is just one params argument 
updates = lasagne.updates.nesterov_momentum( loss, params, learning_rate=0.01, momentum=0.9)
```

If you want to update both at once, just do:
```python
loss = loss_1 + loss_2
params = lasagne.layers.get_all_params([network_1, network_2], trainable=True)
updates = lasagne.updates.nesterov_momentum(loss, params, ...)
```

> ```python
# define train function , but we have two target_var ? 

train= theano.function([input_var, target_var], loss, updates=updates)
```

```python
train = theano.function([input_var, target_var_1, target_var_2], loss, updates=updates)
```
Then pass inputs and both labels for each training example.

For the future, please post such questions on our mailing list; the issue tracker is meant for bug reports and feature discussions only. Thank you!

**Comment by BrightFeather** [comment]
The function runs pretty well. Yet I'd like to know how the output layer should look life. Currently, I got two output of type DenseLayer, how should I get them two, as [out1, out2] does not work, neither do T.stack(out1, out2).
Thanks!

**Comment by f0k** [comment]
> Currently, I got two output of type DenseLayer, how should I get them two, as [out1, out2] does not work, neither do T.stack(out1, out2).

For most purposes, a network with multiple outputs can be represented as a list of all output layers. E.g., if your layers (not output expressions) are `out1` and `out2`, you can pass `[out1, out2]` to `lasagne.layers.get_output` and `lasagne.layers.get_all_params`, as demonstrated above. What is not working in your case?

**Comment by shenjjjj** [comment]
Do we need to scale the different loss functions? Will it be dominated by one of the loss objectives?

**Comment by f0k** [comment]
> Will it be dominated by one of the loss objectives?

This depends on the objectives, but it's common to scale one of the losses (which adds a hyperparameter you need to tune). It can help to monitor both losses separately during training (e.g., have your training function return not only the total loss, but also the two components) so you can see what scales they are and how they develop. This gives you an idea on whether you should boost or weaken the influence of one of the losses over the other.
